docs/proposals/scheduling/20230415-numa-topology-scheduling.md

```diff
@@ -0,0 +1,236 @@
+
+Although Kubernetes’ kubelet provides Topology Manager, it works on the node side and is not aware of the scheduler. This may cause the scheduled Pod to encounter Topology-related errors when it is scheduled to a certain node, resulting in the Pod being unable to start. And this kind of error is sometimes unacceptable. For example, some job-class workloads may cause the entire job to fail due to this reason. Another problem is that it is impossible to obtain cluster-level optimal NUMA Node resources.
+
+Therefore, a mechanism needs to be implemented to solve this requirement, which can coordinate different resource allocation plugins to uniformly align according to NUMA Topology and obtain optimal performance.
+
+# Goal
+
+1. Provide multiple NUMA Topology management strategies to support flexible resource management requirements.
+1. Provide a mechanism to ensure that when scheduling, CPU/Device and other resources are coordinated and the best/preferred NUMA Node at the Pod-Level is obtained.
+1. The scheduler adds extension points to support the collaborative allocation of different resource allocation plugins.
+1. Describe the capabilities that koordlet needs to guarantee for different NUMA Topology strategies.
+
+# Non-Goals
+
+1. There is no limit to the memory management mechanism under the NUMA architecture. Users can set different kernel parameters according to the actual situation to decide whether to allow memory to be accessed across NUMA Nodes.
+
+# User Stories
+
+## Story 1
+
+The user applies for CPU, GPU and RDMA, and expects the allocated resources to be on the same NUMA node to obtain optimal performance.
+
+## Story 2
+
+The user applies for CPU, GPU and RDMA at the same time, and expects the allocated resources to be aligned according to the NUMA node. At this time, it may be necessary to use multiple NUMA nodes.
+
+# Proposal
+
+## Design Detail
+
+### NUMA Topology Policy
+
+The label `node.koordinator.sh/numa-topology-policy` represents that how to aligning resource allocation according to the NUMA topology. 
+- `None` is the default policy and does not perform any topology alignment.
+- `BestEffort` indicates that preferred select NUMA Node that is topology alignment, and if not, continue to allocate resources to Pods.
+- `Restricted` indicates that each resource requested by a Pod on the NUMA Node that is topology alignment, and if not, koord-scheduler will skip the node when scheduling. This strategy requires that all resources be on the same NUMA Node. There can be multiple NUMA Nodes. For example, if a Pod requests 33C and each NUMA Node has 32C, then it can be allocated to use two NUMA Nodes. If this Pod also needs to request GPU/RDMA, then it needs to be on the same NUMA Node as the CPU. This strategy is more flexible.
```

Comment:
what's difference between BestEffort and Restricted?

Reply:
Updated description.

- `BestEffort` indicates that the node *does not strictly* allocate resources according to NUMA Topology alignment. As long as the remaining total amount of nodes meets the needs of Pods, the scheduler can always allocate such nodes to Pods.
- `Restricted` indicates that the node allocates resources *strictly* according to NUMA Topology alignment, that is, the scheduler is required to only select the same one or more NUMA Nodes when allocating multiple resources, otherwise the node should not be used.